### PR TITLE
Support scalar tensors in c#

### LINKED ISSUE #4369: ArgumentException: Dimensions must contain elements
**Describe the bug**
While debugging a network and splitting it into separate parts I reduced one part down to simply returning the first dimension of the shape. This exception occurs while running the model using the C# InferenceSample project. Exception occurs in Tensor.cs:309.

**Urgency**
None

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10 Build 18363
- ONNX Runtime installed from (source or binary): source
- ONNX Runtime version: b41fcf1
- Python version: N/A
- Visual Studio version (if applicable): 2019
- GCC/Compiler version (if compiling from source): cl.exe 19.26.28806 for x64
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

**To Reproduce**
- Use the Microsoft.ML.OnnxRuntime.InferenceSample project
- Modify the sample to use the attached .onnx file (see below) and skip using the inputData from "bench.in" (i.e. initialize the input tensors without inputData)
- Run and notice the exception that occurs.

**Expected behavior**
Since the output is a scalar I'd still expect the dimensions to correctly show as 1.

**Screenshots**
image

**Additional context**
[siamese_test_opt.zip](https://github.com/microsoft/onnxruntime/files/4848364/siamese_test_opt.zip)

### PR DESCRIPTION
**Pull request by hariharans29**
**Description**: The C# layer currently doesn't support creating scalars which cause issues in models that do require scalars to be created. This change adds support for scalars in C#.

Refined dupe of #4394 (easier to iterate in a local branch than on a fork)

Contribution credit: @amirebrahimi

**Motivation and Context**
Fixes #4369

### CODE REVIEW

File: csharp/test/Microsoft.ML.OnnxRuntime.Tests/Tensors/TensorTests.cs [hunk 1]
```diff
@@ -160,7 +160,6 @@ public void ConstructFromDimensions(TensorConstructor tensorConstructor)
             Assert.Equal(tensorConstructor.IsReversedStride, tensor.IsReversedStride);
 
             //Assert.Throws<ArgumentNullException>("dimensions", () => tensorConstructor.CreateFromDimensions<int>(dimensions: null));
-            Assert.Throws<ArgumentException>("dimensions", () => tensorConstructor.CreateFromDimensions<int>(dimensions: new int[0]));
```
Comment (hariharans29) [inline csharp/test/Microsoft.ML.OnnxRuntime.Tests/Tensors/TensorTests.cs:163]: This test is not valid anymore - in place of this - we have extensive tests covering multiple aspects of scalar tensors below.

File: csharp/test/Microsoft.ML.OnnxRuntime.Tests/Tensors/TensorTests.cs [hunk 2]
```diff
@@ -224,12 +223,88 @@ public void ConstructEmptyTensors(TensorConstructor tensorConstructor)
...
-            var emptyDenseTensor = new DenseTensor<int>(new Span<int>(dimensions));
+            var emptyDenseTensor1 = new DenseTensor<int>(new Span<int>(dimensions));
+            Assert.Equal(0, emptyDenseTensor1.Length);
             // accessing any index in the underlying buffer should result in an IndexOutOfRangeException
-            Assert.Throws<IndexOutOfRangeException>(() => emptyDenseTensor.GetValue(0));
-            Assert.Throws<IndexOutOfRangeException>(() => emptyDenseTensor.GetValue(5));
+            Assert.Throws<IndexOutOfRangeException>(() => emptyDenseTensor1.GetValue(0));
+            Assert.Throws<IndexOutOfRangeException>(() => emptyDenseTensor1.GetValue(5));
+
+            // create an empty DenseTensor from memory
```
Comment (hariharans29) [inline csharp/test/Microsoft.ML.OnnxRuntime.Tests/Tensors/TensorTests.cs:234]: for completeness' sake - adding this test for "empty" tensor as well

File: csharp/src/Microsoft.ML.OnnxRuntime/Tensors/Tensor.cs [hunk 3]
```diff
@@ -397,15 +397,19 @@ protected Tensor(int length) : base(typeof(T))
...
+        /// Initialize an n-dimensional tensor with the specified dimensions and layout.  ReverseStride=true gives a stride of 1-element width to the first dimension (0).  ReverseStride=false gives a stride of 1-element width to the last dimension (n-1).
         /// </summary>
         /// <param name="dimensions">An span of integers that represent the size of each dimension of the Tensor to create.</param>
         /// <param name="reverseStride">False (default) to indicate that the first dimension is most major (farthest apart) and the last dimension is most minor (closest together): akin to row-major in a rank-2 tensor.  True to indicate that the last dimension is most major (farthest apart) and the first dimension is most minor (closest together): akin to column-major in a rank-2 tensor.</param>
         protected Tensor(ReadOnlySpan<int> dimensions, bool reverseStride) : base(typeof(T))
         {
+            // Scalar tensor
             if (dimensions.Length == 0)
```
Comment (yuslepukhin) [inline csharp/src/Microsoft.ML.OnnxRuntime/Tensors/Tensor.cs:407]: > if (dimensions.Length == 0) [](start = 11, length = 28)

Should we add else clause to it? Everything done here is overwritten below

File: csharp/src/Microsoft.ML.OnnxRuntime/Tensors/Tensor.cs [hunk 4]
```diff
@@ -438,9 +442,15 @@ protected Tensor(Array fromArray, bool reverseStride) : base(typeof(T))
                 throw new ArgumentNullException(nameof(fromArray));
             }
 
+
+            // Scalar tensor
+            // TODO: Check if it is possible to create a zero-ranked array ?
             if (fromArray.Rank == 0)
```
Comment (yuslepukhin) [inline csharp/src/Microsoft.ML.OnnxRuntime/Tensors/Tensor.cs:448]: >if (fromArray.Rank == 0) [](start = 12, length = 24)

else clause?

File: csharp/test/Microsoft.ML.OnnxRuntime.Tests/Tensors/TensorTests.cs [hunk 5]
```diff
@@ -160,7 +160,6 @@ public void ConstructFromDimensions(TensorConstructor tensorConstructor)
             Assert.Equal(tensorConstructor.IsReversedStride, tensor.IsReversedStride);
 
             //Assert.Throws<ArgumentNullException>("dimensions", () => tensorConstructor.CreateFromDimensions<int>(dimensions: null));
```
Comment (yuslepukhin) [inline csharp/test/Microsoft.ML.OnnxRuntime.Tests/Tensors/TensorTests.cs:162]: > //Assert.Th [](start = 11, length = 12)

Remove commented code  if not needed

File: csharp/src/Microsoft.ML.OnnxRuntime/Tensors/Tensor.cs (hunk 3 above)
Comment (hariharans29) [reply csharp/src/Microsoft.ML.OnnxRuntime/Tensors/Tensor.cs:407]: Oh yeah. looks like I forgot to copy over the else when I copied this piece of code from the original PR Actually - we may consider dropping this code altogether - looks like what exists can handle scalars - no need to special case it.

Comment (hariharans29) [reply csharp/src/Microsoft.ML.OnnxRuntime/Tensors/Tensor.cs]: Took out the entire "special case" logic. The existing code can handle it as such. Just removed the throwing exception part.

File: csharp/src/Microsoft.ML.OnnxRuntime/Tensors/Tensor.cs (hunk 4 above)
Comment (hariharans29) [reply csharp/src/Microsoft.ML.OnnxRuntime/Tensors/Tensor.cs]: Ditto

File: csharp/test/Microsoft.ML.OnnxRuntime.Tests/Tensors/TensorTests.cs (hunk 5 above)
Comment (hariharans29) [reply csharp/test/Microsoft.ML.OnnxRuntime.Tests/Tensors/TensorTests.cs]: Actually kept the test - looks like this was checked-in by mistake. The code guarding against null dimensions and this test should both exist.
